dhis-2/dhis-api/src/main/java/org/hisp/dhis/dataintegrity/FlattenedDataIntegrityReport.java

```diff
@@ -151,151 +155,144 @@
 
-        programRuleActionsWithNoDataObject = transformMapOfCollections(
+        programRuleActionsWithNoDataObject = mapToDisplayNameOrId(
             report.getProgramRuleActionsWithNoDataObject() );
 
-        programRuleActionsWithNoNotification = transformMapOfCollections(
+        programRuleActionsWithNoNotification = mapToDisplayNameOrId(
             report.getProgramRuleActionsWithNoNotification() );
 
-        programRuleActionsWithNoSectionId = transformMapOfCollections( report.getProgramRuleActionsWithNoSectionId() );
+        programRuleActionsWithNoSectionId = mapToDisplayNameOrId( report.getProgramRuleActionsWithNoSectionId() );
 
-        programRuleActionsWithNoStageId = transformMapOfCollections( report.getProgramRuleActionsWithNoStageId() );
+        programRuleActionsWithNoStageId = mapToDisplayNameOrId( report.getProgramRuleActionsWithNoStageId() );
     }
 
     // -------------------------------------------------------------------------
     // Supportive methods
     // -------------------------------------------------------------------------
 
-    private Collection<Collection<String>> transformCollectionOfCollections(
+    private List<List<String>> mapToListOfListOfDisplayNameOrUid(
         Collection<? extends Collection<? extends IdentifiableObject>> collection )
     {
-        Collection<Collection<String>> newCollection = new HashSet<>();
-
-        for ( Collection<? extends IdentifiableObject> c : collection )
-        {
-            newCollection.add( transformCollection( c ) );
-        }
-
-        return newCollection;
+        return collection == null
+            ? null
+            : collection.stream().map( this::mapToListOfDisplayNameOrUid ).collect( toUnmodifiableList() );
     }
 
-    private Map<String, String> transformMapOfStrings( Map<? extends IdentifiableObject, String> map )
+    private Map<String, String> mapKeyToDisplayNameOrUid( Map<? extends IdentifiableObject, String> map )
     {
-        HashMap<String, String> newMap = new HashMap<>( map.size() );
-
-        for ( Map.Entry<? extends IdentifiableObject, String> entry : map.entrySet() )
-        {
-            newMap.put( defaultIfNull( entry.getKey() ), entry.getValue() );
-        }
-
-        return newMap;
+        return map == null
+            ? null
+            : map.entrySet().stream().collect( toUnmodifiableMap(
+                e -> defaultIfNull( e.getKey() ), Entry::getValue ) );
     }
 
-    private Map<String, Collection<String>> transformMapOfCollections(
+    private Map<String, Collection<String>> mapToDisplayNameOrId(
         Map<? extends IdentifiableObject, ? extends Collection<? extends IdentifiableObject>> map )
     {
-        HashMap<String, Collection<String>> newMap = new HashMap<>();
-
-        for ( var entry : map.entrySet() )
-        {
-            newMap.put( defaultIfNull( entry.getKey() ), new HashSet<>( transformCollection( entry.getValue() ) ) );
-        }
-
-        return newMap;
```

Comment:
FYI: I could not see any meaningful semantic in mapping to hash maps and sets order by hash code. This basically gives a random order. So this and the other "flattening" helper do not maintain the exact same semantic when it comes to order since this was randon as far as I can tell (except where the sorted map was used which it still is).